From f103aa6fd13dfa676f1b3646d4dcaee82f97b3a8 Mon Sep 17 00:00:00 2001
From: Peter Vary <pvary@cloudera.com>
Date: Tue, 15 Aug 2017 18:04:26 +0200
Subject: [PATCH 1247/1363] CDH-58028 HIVE-17268: WebUI / QueryPlan: query
 plan is sometimes null when explain output conf
 is on (Karen Coppage via Peter Vary)

(cherry picked from commit d3b820f641355b2bc355c5be5b76f38b028edc2c)

Change-Id: I5f5ae934fe24a6f542556f1c411bce55b6c7f980
---
 .../hive/service/cli/session/TestQueryDisplay.java |    3 ++-
 .../org/apache/hadoop/hive/ql/QueryDisplay.java    |    2 +-
 .../org/apache/hive/tmpl/QueryProfileTmpl.jamon    |   14 +++++++++++---
 .../hive/service/servlet/QueryProfileServlet.java  |    4 +++-
 4 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/itests/hive-unit/src/test/java/org/apache/hive/service/cli/session/TestQueryDisplay.java b/itests/hive-unit/src/test/java/org/apache/hive/service/cli/session/TestQueryDisplay.java
index 06b4af5..8be9d54 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/service/cli/session/TestQueryDisplay.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/service/cli/session/TestQueryDisplay.java
@@ -197,7 +197,8 @@ private void verifyDDLHtml(String stmt, String opHandle) throws Exception {
     StringWriter sw = new StringWriter();
     QueryInfo queryInfo = sessionManager.getOperationManager().getQueryInfo(
       opHandle);
-    new QueryProfileTmpl().render(sw, queryInfo);
+    HiveConf hiveConf = sessionManager.getOperationManager().getHiveConf();
+    new QueryProfileTmpl().render(sw, queryInfo, hiveConf);
     String html = sw.toString();
 
     Assert.assertTrue(html.contains(stmt));
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/QueryDisplay.java b/ql/src/java/org/apache/hadoop/hive/ql/QueryDisplay.java
index 338366c..467dab6 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/QueryDisplay.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/QueryDisplay.java
@@ -189,7 +189,7 @@ public synchronized String getQueryString() {
   }
 
   public synchronized String getExplainPlan() {
-    return explainPlan == null ? "SET hive.log.explain.output TO true TO VIEW PLANS" : explainPlan;
+    return returnStringOrUnknown(explainPlan);
   }
 
   public synchronized void setExplainPlan(String explainPlan) {
diff --git a/service/src/jamon/org/apache/hive/tmpl/QueryProfileTmpl.jamon b/service/src/jamon/org/apache/hive/tmpl/QueryProfileTmpl.jamon
index cbd2f8b..b7ede8b 100644
--- a/service/src/jamon/org/apache/hive/tmpl/QueryProfileTmpl.jamon
+++ b/service/src/jamon/org/apache/hive/tmpl/QueryProfileTmpl.jamon
@@ -18,11 +18,13 @@ limitations under the License.
 </%doc>
 <%args>
 QueryInfo queryInfo;
+HiveConf hiveConf;
 </%args>
 <%import>
 java.util.*;
 org.apache.hadoop.hive.ql.QueryDisplay;
 org.apache.hadoop.hive.ql.QueryInfo;
+org.apache.hadoop.hive.conf.HiveConf;
 </%import>
 <!--[if IE]>
 <!DOCTYPE html>
@@ -204,9 +206,15 @@ org.apache.hadoop.hive.ql.QueryInfo;
     <div class="panel panel-default">
       <div class="panel-heading">Explain plan</div>
       <div class="panel-body">
-        <pre>
-        <% queryInfo.getQueryDisplay() == null ? "Unknown" : queryInfo.getQueryDisplay().getExplainPlan() %>
-        </pre>
+        <%if hiveConf.getBoolVar(HiveConf.ConfVars.HIVE_LOG_EXPLAIN_OUTPUT) %>
+          <pre>
+          <% queryInfo.getQueryDisplay() == null ? "Unknown" : queryInfo.getQueryDisplay().getExplainPlan() %>
+          </pre>
+        <%else>
+          <pre>
+          Set configuration hive.log.explain.output to true to view future query plans
+          </pre>
+        </%if>
       </div>
     </div>
 </%def>
diff --git a/service/src/java/org/apache/hive/service/servlet/QueryProfileServlet.java b/service/src/java/org/apache/hive/service/servlet/QueryProfileServlet.java
index 27a3b14..4996374 100644
--- a/service/src/java/org/apache/hive/service/servlet/QueryProfileServlet.java
+++ b/service/src/java/org/apache/hive/service/servlet/QueryProfileServlet.java
@@ -17,6 +17,7 @@
  */
 package org.apache.hive.service.servlet;
 
+import org.apache.hadoop.hive.conf.HiveConf;
 import org.apache.hadoop.hive.ql.QueryInfo;
 import org.apache.hive.service.cli.operation.OperationManager;
 import org.apache.hive.service.cli.session.SessionManager;
@@ -47,11 +48,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
       (SessionManager)ctx.getAttribute("hive.sm");
     OperationManager opManager = sessionManager.getOperationManager();
     QueryInfo queryInfo = opManager.getQueryInfo(opId);
+    HiveConf hiveConf = opManager.getHiveConf();
     if (queryInfo == null) {
       LOG.debug("No display object found for operation {} ", opId);
       return;
     }
 
-    new QueryProfileTmpl().render(response.getWriter(), queryInfo);
+    new QueryProfileTmpl().render(response.getWriter(), queryInfo, hiveConf);
   }
 }
-- 
1.7.9.5

